1. Observe that the regularized problem (2) can be written as
min
x
fλ(x) = min
x
XN
i=1
fi(x) (3)
Find an appropriate choice of fi(x).

Answer-1

We can write f(x) in the  form of $f_i(x)$, we can let:
$$
f_i(x)=\frac{1}{2}\left\|a_i^T x-y_i\right\|^2+\frac{\lambda}{2N} x^Tx
$$

$ min_x f_\lambda(x)=min_x\sum_{i=1}f_i(x)$


so,

Gradient will be

$$
 \nabla{f_i(x)}= \sum_{i=1}^n(a_i^T x-y_i)a_i + \lambda x
$$

2. Write an expression to compute the gradient of fi(x) and denote it by gi(x) = ∇xfi(x).

answer-2

We already calculated gradient above, so expression to compute the gradient of fi(x) and denote it by is-

$$
g_i(x) = \nabla{f_i(x)}=\nabla_x{ ( \frac{1}{2}\left\|a_i^T x-y_i\right\|^2+\frac{\lambda}{n} x^Tx )}
$$
$$
 = (a_i^Tx - y_i)a_i + \frac{\lambda}{n}x
$$

3. Consider the dimension where you observed failure in the previous exercise. Implement the following algorithm
(ALG-LAB6) to solve (3):

In [1]:
import numpy as np
from numpy.linalg import norm
import time
import timeit

np.random.seed(1000) #for repeatability
N = 200
d = 20000 #in the last queztion failue occure ar 20000
lambda_reg = 0.001
eps = np.random.randn(N,1)
A = np.random.randn(N,d)
#Normalize the columns
for j in range(A.shape[1]):
  A[:,j] = A[:,j]/np.linalg.norm(A[:,j])
xorig = np.ones( (d,1) )
y = np.dot(A,xorig) + eps


In [2]:
def fx(x, lamda):
  return (1/2)*norm(A@x-y)**2 + (1/2)*lamda*np.dot(x,x)

def grad_fx(x, lamda):
  sum = np.array([0. for _ in range(d)])
  for i in range(N):
    sum = sum+(A[i]@x - y[i])[0]*A[i]
  sum =sum +  lamda*x
  return sum

In [7]:
x = np.zeros((d,1))
x = x.flatten()
epochs = 10**4
t = 1
arr = np.arange(N)
start = timeit.default_timer()
for epoch in range(epochs):
  np.random.shuffle(arr)
  for i in np.nditer(arr):
    gi = (A[i]@x - y[i])[0]*A[i] + (lambda_reg/N)*x
    x = x - (gi/t)
    # Update x using x <- x- 1/t * g_i (x)
    t = t+1
    if t>1e4:
      t = 1
algtime = timeit.default_timer()- start #time is in seconds
x_opt = x
print("Total Epochs is ", epochs)
print("Time Taken is ", algtime)
print("Norm of Gradient at x opt is", norm(grad_fx(x_opt, lambda_reg)))
print("||Ax_alg- y||^2  is",norm(A@x - y)**2)
print("||x_opt- xorig||^2 is   ", norm(x_opt - xorig.flatten())**2)

Total Epochs is  10000
Time Taken is  350.0029364310001
Norm of Gradient at x opt is 0.014280438515986649
||Ax_alg- y||^2  is 6274940.964338467
||x_opt- xorig||^2 is    19847.9190882912


**Report**

Time taken= 350.00293643100015 sec

Norm of gradient at x_opt = 0.014280438515986649

||Ax_alg- y||^2 = 6274940.964338467

||x_opt- xorig||^2 :    19847.919033137703

4. Fix λ = 0.001 and repeat the experiment for 106, 108 amd 1010 epochs and report the time taken for ALG-LAB6,
∥∇fλ(x∗)∥, ∥Ax∗ − y∥22
and ∥x∗ − xorig∥22
. Explain your observations.

In [8]:
epc = np.array([10**3, 10**5], dtype = int)
for epoc in epc:
  x = np.zeros((d,1))
  x = x.flatten()
  t = 1
  arr = np.arange(N)
  start = timeit.default_timer()
  for epoch in range(epoc):
    # print("epoch no.: ", epoch)
    np.random.shuffle(arr) #shuffle every epoch
    for i in np.nditer(arr): #Pass through the data points
      gi = (A[i]@x - y[i])[0]*A[i] + (lambda_reg/N)*x
      x = x - (gi/t)
      # Update x using x <- x- 1/t * g_i (x)
      t = t+1
      if t>1e4:
        t = 1
  algtime = timeit.default_timer()- start #time is in seconds
  x_opt = x
  print("Total Epochs: ", epoc)
  print("Time Taken: ", algtime)
  print("Norm of Gradient at x-opt : ", norm(grad_fx(x_opt, lambda_reg)))
  print("||Ax_alglab6- y||^2  : ",norm(A@x - y)**2)
  print("||x_opt- xorig||^2 :   ", norm(x_opt - xorig.flatten())**2)
  #print the time taken, ||Ax_alglab6- y||^2, ||x_opt- xorig||^2

Total Epochs:  1000
Time Taken:  37.17554169899995
Norm of Gradient at x-opt :  0.01647208488895905
||Ax_alglab6- y||^2  :  6274945.118619619
||x_opt- xorig||^2 :    19847.919066320854
Total Epochs:  100000
Time Taken:  3434.3294305740005
Norm of Gradient at x-opt :  0.01990010104843626
||Ax_alglab6- y||^2  :  6274946.206369429
||x_opt- xorig||^2 :    19847.919111883606


**Observation:**

Increasing the total number of epochs from 1000 to 100000 significantly increases the time taken for optimization, from 37.18 seconds to 3434.33 seconds, indicating a substantial increase in computational cost with more epochs.

The norm of the gradient at the optimal solution slightly increases from 0.0165 to 0.0199 when the number of epochs increases, indicating a potentially slower convergence rate with a larger number of epochs.

The values of ||Ax_alglab6- y||^2 and ||x_opt- xorig||^2 remain almost unchanged between the two cases, suggesting that the difference between the predicted output and actual output, as well as the difference between the optimal solution and the original solution, is not significantly affected by the number of epochs beyond 1000.

**NOTE:** I have taken only $10^3,10^5$ of epoch values becuase TA's tell us to take these values

5. Fix 109 and try λ ∈ {1000, 100, 10, 1, 0.1, 10−2, 10−3} and report the time taken for ALG-LAB6 ∥∇fλ(x∗)∥,
∥Ax∗ − y∥22
and ∥x∗ − xorig∥22
. Explain your observations.

In [10]:
lamdas = [10**3, 10**2, 10, 1, 0.1, 0.01, 0.001]
epochs = 10**4

for lamda in lamdas:
  x = np.zeros((d,1))
  x = x.flatten()
  t = 1
  arr = np.arange(N) #index array
  start = timeit.default_timer() #start the timer
  for epoch in range(epochs):
    # print("epoch no.: ", epoch)
    np.random.shuffle(arr) #shuffle every epoch
    for i in np.nditer(arr): #Pass through the data points
      gi = (A[i]@x - y[i])[0]*A[i] + (lamda/N)*x
      x = x - (gi/t)
      # Update x using x <- x- 1/t * g_i (x)
      t = t+1
      if t>1e4:
        t = 1

  algtime = timeit.default_timer()- start #time is in seconds
  x_opt = x
  print("Total Epochs: is ", epochs)
  print("Lambda Regularizer taken is ", lamda)
  print("Time Taken is  ", algtime)
  print("Norm of Gradient at x_opt is ", norm(grad_fx(x_opt, lamda)))
  print("||Ax_alg- y||^2  is ",norm(A@x - y)**2)
  print("||x_opt- xorig||^2 is   ", norm(x_opt - xorig.flatten())**2)

Total Epochs: is  10000
Lambda Regularizer taken is  1000
Time Taken is   377.43905466600154
Norm of Gradient at x_opt is  11.030815456470403
||Ax_alg- y||^2  is  3168743.0548190055
||x_opt- xorig||^2 is    19972.932410778765
Total Epochs: is  10000
Lambda Regularizer taken is  100
Time Taken is   355.99423169799957
Norm of Gradient at x_opt is  37.008674573943516
||Ax_alg- y||^2  is  3974633.1820549257
||x_opt- xorig||^2 is    19883.18775236549
Total Epochs: is  10000
Lambda Regularizer taken is  10
Time Taken is   408.64287357600006
Norm of Gradient at x_opt is  80.24782693457189
||Ax_alg- y||^2  is  5837626.342537512
||x_opt- xorig||^2 is    19848.752893582514
Total Epochs: is  10000
Lambda Regularizer taken is  1
Time Taken is   347.7240635440012
Norm of Gradient at x_opt is  15.675638760255639
||Ax_alg- y||^2  is  6235253.438350475
||x_opt- xorig||^2 is    19847.955216602193
Total Epochs: is  10000
Lambda Regularizer taken is  0.1
Time Taken is   342.2555698979995
Norm of Gradient

Observing the results, as the regularization parameter λ decreases, the norm of the gradient at the optimum also decreases, indicating better convergence. However, reducing λ too much can lead to overfitting, as seen in the increase of ||Ax_alg- y||^2. The time taken generally increases with smaller λ due to more iterations needed for convergence. Balancing between regularization strength, convergence speed, and model fit is crucial. Despite differences in regularization, the final ||x_opt- xorig||^2 remains relatively stable, suggesting the regularization impact on the solution's closeness to the original.

6. Does ALG-LAB6 work for failure dimension?

**Answer:**  Yes alg-Lab6 works for the failure dimention of last question.In the last question failure occure at 20000 so I have only check at that point but not for large values of d

7. Explain your understanding of ALG-LAB6.

**Observations:**

ALG-LAB6 appears to be a stochastic optimization algorithm, likely a variant of stochastic gradient descent (SGD) due to its per-sample update rule.

It aims to minimize the difference between the predicted output Ax_alglab6 and the actual output y by updating the optimization variable x iteratively.
The algorithm employs a learning rate schedule where the learning rate decreases over time as 1/t, where t is the iteration counter.

The shuffling of data points in each epoch helps in preventing the algorithm from getting stuck in local minima and aids in better exploration of the solution space.

The reported metrics such as time taken, squared norm of residual, and squared norm of difference provide insights into the performance and accuracy of the algorithm.